### ISSUE TITLE
`Identity`type really needed? (Probably reduce allocations / usage)

### ISSUE THREAD

**Issue by kellertuer** [JuliaManifolds/Manifolds.jl#378]
In relation to #376 I stumbled upon our (quite a few) ambiguities concerning operators and identities, so I asked a question on discourse (https://discourse.julialang.org/t/avoid-ambiguities-with-individual-number-element-identity/) which made me wonder - why did we introduce that type in the first place? Do we really need it in some dispatch, where we can not get around the extra type?

I feel that just getting the matrix representation of the identity, we could use `one(M,x)` or `one(M)` even (or stick with `identity(M)` since it would only be a `one` for the multiplicative action).

Any thoughts on this @sethaxen ? Maybe I missed where we really need this type.

**Comment by mateuszbaran** [comment]
I guess `Identity` enables some optimizations but the benefit seems very small compared to the cost of maintaining that type. I guess we could just ignore any problems with `Identity` (filter our any ambiguities where `Identity` appears in signature of either method) until someone complains. That would be the lazy non-breaking solution.

**Comment by kellertuer** [comment]
Yes, that seems to be a good solution.

**Comment by mateuszbaran** [comment]
This seems to filter out `Identity`:
```julia
function has_type_in_sig(m, T::Type)
    return any(map(Base.unwrap_unionall(m.sig).parameters) do x
        xw = Base.rewrap_unionall(x, m.sig)
        return xw <: T
    end)
end

using Manifolds, Test

ms = Test.detect_ambiguities(Manifolds);

fms = filter(x -> !any(has_type_in_sig.(x, Identity)), ms)
```

**Comment by kellertuer** [comment]
I will work on this in #377 anyways, so this can basically be closed, since Identity seems to be needed, we just have to take care of some ambiguities.

**Comment by kellertuer** [comment]
I would like to reopen this; @sethaxen since you introduced the Identity type, we have had a short discussion over on [discourse](https://discourse.julialang.org/t/avoid-ambiguities-with-individual-number-element-identity/62465/) that it might be better (especially for the about 140 ambiguities it introduces) to get rid of the identity type. What do you think?

**Comment by sethaxen** [comment]
When `Identity` was originally introduced, it did not store a point, and so it was an allocation-free way to signal to the compiler that some operations were no-ops and should not be computed. But now that it stores a point, I'm not sure it's worth it anymore, especially given the complexity that you mentioned.

**Comment by kellertuer** [comment]
Thanks for the feedback.

Having an allocation free identity might be nice, for sure; currently it introduces about 150 ambiguities, from which I am not so sure they are harmless – and you are right, if we now store the data anyways that is not useful anymore.
I wasn't aware of your intention, I am not 100% sure it would spare that much memory allocations, but in the current state, I would propose to remove the identity type again? I can surely do that, though it might take a little while until I find time

**Comment by mateuszbaran** [comment]
It looks like at some point it was decided to allow `Identity` at any place where a point on a Lie group is allowed. Which may have some use because sometimes it makes it possible to write faster functions, but it's rarely used that way.

Also, ambiguities are perfectly harmless as long as `Identity` isn't used in the wrong place. `Identity` has good use cases but using it correctly seems to be quite hard.

**Comment by kellertuer** [comment]
It seems with the new Connections you sometimes dispatch on identity. So do we keep Identity or not?
For me it seems to be not so nice if we store point and group in the identity. If we do not do that (i.e. avoid memory and allocations) I see that it might be useful,.

removing Identity, I am not 100% sure what to do for the vector transports that specialise on this currently. For the rest it is just removing those cases.

**Comment by kellertuer** [comment]
Maybe a good way in between is – keep it for the allocation free motivation but do not store point/manifold in there anymore. This will not make the type applicable everywhere, but maybe reduce its complexity? Also we currently nee a lot of lines of codes for functions to not have ambiguities with identity; also our own ones. Maybe that would be an argument to really remove it?

**Comment by mateuszbaran** [comment]
Yes, I agree that it'd be better to not store the point in `Identity`. I think the point is there mostly as a fallback option.

I think `Identity` is most problematic as an argument of `get_vector` and `get_coordinates`, so we could start with removing it from there?

**Comment by kellertuer** [comment]
I think so, too. There it does not provide much information and/or help. I will check how that can best be done, so we end up with a simpler `Identity` (without fields).

**Comment by kellertuer** [comment]
I today tried to work a little bit on this and ran into an interesting point

If we want to make Identity similar to our points, it should also not have the group in its type. So I tried to refactor even to `Identity` (not just `Identity{<:AbstractGroupManifold}`) but I think we get into trouble with that, since the overloaded operators `+`,`-`,`*`,`/` then introduce strange side effects, since for `e=Identity()`

`p+e` will always just be `p` also when `p` belongs to an `MultiplicationGroup` where `get_point(G,e)` would be `1`. On the other hand, I am not sure `+` is a suitable operation on an multiplicative group in general.

In general I like the idea that `Identity` should be as agnostic of the manifold as any point `p` is.
But we could also type it per Groupmanifold, if needed.

What do we do about that?

**Comment by mateuszbaran** [comment]
Interesting. What do you think about making `Identity{<:AbstractGroupOperation}`? That would make overloaded operators less awkward to use with `Identity`.

**Comment by kellertuer** [comment]
That might be an idea. I will first try to finish the untyped one to see where that ends up – and then try how much changes if we type by operation; but for me the operation idea sounds quite reasonable.

**Comment by kellertuer** [comment]
So, I fiddled around quite some time yesterday and here is my findings and my proposal what to do

1. For the operations to always work fine, we can not just use a parameter-free `Identity`. @mateuszbaran proposed to add the group action, and I think that might be neccessary, but it also introduces the necessity to handle incompatible indentities here and there, which is – in my opinion just chasing special cases and throw errors then.

2. The one thing that troubled me yesterday and still is, is the following: If `Identity` is allocation free and still should act “point-like” we have some areas where we run into areas, where we need to call `get_point(G,e)` to get a point representation (i.e. array) of the `Identity e`. For that `e` has to know which type points are (to produce an `Array` identity or an `SArray` identity, including the right `eltype`). So I think we can not get around this if we keep this.

If we keep `Identity`, then I think we have to do 2 and we should do 1, too. I am not sure of the order of parameters then, but the even more interesting question – I think – is: What are the benefits of having an identity type? We spare some few allocations?
Since we have to implement tricks `get_points` for the identity anyways, I would be in favour of just doing

3. instead of `get_point(G,e)` lets implement just `one(G,p)` where `one(G)` would produce a `Array{Float64}` or `Array{ComplexxF64}` (maybe even a quaternion one? we do not have those in our dependencies yet) depending on the field type. This would always produce a array-version of the identity element of `G` and would spare all the hustle with identity, both in the current variant, in (1&)2, as well as reduce ambiguities. I am not sure keeping Identity is worth the reduced allocations.

I would be in favour of 3, since it is easy, reduces complexity and is close to a Julia-like style (compared to the current that does not spare allocations either). If we have an allocation-free Identity, then we have to do 2, and maybe also 1.

What is your opinion on this? 1&2 _or_ 3?

**Comment by mateuszbaran** [comment]
I can try doing (1) myself, and if that's too complicated we can just drop it.

Regarding (2) and (3), as I wrote in a comment to the PR, why do we need `get_point` with `Identity` as an argument? Do you actually mean the `get_point` related to charts?

**Comment by kellertuer** [comment]
I feel that (1) would be quite technical and complicated, so the question is, is it really worth it? I feel like we introduce a lot technicalities here and I am not sure what for. As the discourse thread proposed, maybe it is simpler to drop this type. We have a lot of technicalities with it to spare the memory of a few bytes.

No I did not eat it in the sense of charts, but `get_array` also did not fit since it might also be an `AbstractManifoldPoint`, I just did not find a better name

**Comment by mateuszbaran** [comment]
I think something like `get_identity(G)` would be perfectly fine. It doesn't even have to take `e::Identity` as an argument, and can return a point of some default type depending on the group. If someone doesn't like that default, they can always change it.

**Comment by kellertuer** [comment]
Then we can also use `Base.one(G)` where optimally you can provide a type `Base.one(G,p)` to provide the type (a similar point) for the result. Still – this would basically make `Identity` obsolete and would be nicer I think (see my point 3).

**Comment by mateuszbaran** [comment]
One thing I don't like about overloading `Base.one` is that for the group of translation it would return an array of zeros, which might be confusing. Also I don't think it would make the `Identity` type entirely obsolete, just very limited. I think for example the use of `Identity` in `vector_transport_direction` for `CartanSchoutenZeroGroup` is a great example how it should be used.

**Comment by kellertuer** [comment]
Hm, can you propose something on the branch I did? Because I got really lost yesterday trying to get 1&2 to work and that seems really a lot a lot of code for no real use (at least to me). In my opinion, we could really just introduce the three methods (sure, not necessarily `one`see my edit above) and leave `Identity` out. It introduces so much trouble for not much effect.

**Comment by mateuszbaran** [comment]
I can try something next week. In any case I'm fine with removing most of the current usage of `Identity`.

**Comment by kellertuer** [comment]
Ok, then I can help for sure, but I am not sure which parts you want to keep. For example I would not overload `+` and such. I will try to find time to do the new `identity(G,e)` style functions today.

**Comment by kellertuer** [comment]
This was now resolved by #401 with a memory reduced identity and less ambiguities.
